VERSION-RANGE-SPEC.rst

```diff
@@ -0,0 +1,537 @@
+
+Each ``<version-constraint>`` in the comma-separated list is either a simple
+constraint such as::
+
+    <comparator:version>
+
+Or a composite constraint grouping multiple ``<version-constraint>`` joined by
+an ampersand such as:
+
+    <comparator:version>&<comparator:version>...
+
+The comma is a logical `OR` and the ampersand is a logical `AND`.
+
+A version range specifier is therefore an "OR of ANDs". This is also called a
+"disjunctive normal form" in boolean logic.
+See https://en.wikipedia.org/wiki/Disjunctive_normal_form for details.
+
+``vers`` is the URI-scheme and is an acronym for "VErsion Range Specifier". It
+has been selected because it is short, obviously about version and available
+for a future formal registration for this URI-scheme at the IANA registry.
+
+
+``<versioning-scheme>``
+------------------------
+
+The ``<versioning-scheme>`` (such as ``semver``,
+``debian``, etc.) determines:
+
+- the specific notation and conventions used for a version string encoded in
+  this scheme. Versioning schemes often specify a version segments separator and
+  the meaning of each version segments, such as [major.minor.patch] in semver.
+
+- how two versions compare as greater or lesser to determine if a version is
+  within or outside a range.
+
+- how a versioning scheme-specific range notation can be transformed in the
+  ``vers`` notation defined here.
+
+Note: the ``<versioning-scheme>`` does not define which symbols to use to
+compare versions, and how to compose these as this is the substance of this
+specification.
+
+The ``<versioning-scheme>`` is followed by a slash "/".
+
+Then there are one or more ``<version-constraint>`` separated by a comma ","
+meaning that **any** of these constraints must be satisfied for a version to be
+resolved as within this range specifier.
+
+
+``<version-constraint>``
+----------------------------
+
+Each  ``<version-constraint>`` of this comma-separated list can be either a
+single constraint or a list of constraints separated by an ampersand "&" as in
+``1.2.3,>=2.0.0&<5.0.0``.
+
+Multiple ``<version-constraint>`` combined with an ampersand means that **all**
+these constraints must be satisfied for a version to be resolved as within this
+range.
+
+Each simple version constraint has this syntax::
+
+    <comparator><version>
+
+The ``<comparator>`` is one of these comparison operators:
+
+- "=": Version equality comparator. It is the default and implied if not
+  present and means that a version must be equal to the provided version.
+  For example: "=1.2.3". It must be omitted in the canonical representation.
+  Equality is based on the equality of two lower-cased and normalized version
+  strings and is not versioning scheme-specific.
+
+- "!=": Version exclusion or inequality comparator. This means a version must
+  not be equal to the provided version and this version must be excluded from
+  the range. For example: "!=1.2.3" means that version "1.2.3" is not part of
+  the range.
```

Comment:
> Eg: "!=1.2.3, >1.0.0" implies versions should be "greater than 1.0.0 or not equal to 1.2.3" which will consist a set of every possible version.

yes, but you may want to write instead "!=1.2.3 & >1.0.0" which means what you intended: ""greater than 1.0.0 and not equal to 1.2.3" 

Would you have a an example of something that cannot be expressed with this notation?

Reply:
Could it be easier to read a "vers" if this is in CNF, e.g., an AND of ORs?
.. possibly using a coma for AND (like is done in pypi) and a pipe for OR?

- AND is the default for pypi (there is no OR) with a coma; same for debian, rpm and rust: AND is the first thing, and is a coma
- same for node-semver and composer where AND is the default

With that said, the two use cases may be conflicting:

- for vulnerabilities, it is common to have a the same vulnerability affect multiple branches, such as httpd 1.x and 2.x. Here a version that's within any of these ranges would be vulnerable. e.g. any woudl mean an OR first.

- for dependencies, we are trying to resolve and pick one version and in the common case, all the constrainst would need to be satisfied. Some notations do not even support an OR.


So vulnerabilities, OR is first. Dependencies AND is first.

In order to adopt the most common notations in use, it may be better to use AND first as this would be more natural.
But this would mean that for vulnerabilities multiple "vers" might need to be provided, essentially one for each "branch"; it not seem completely crazy but I am not sure I like it.


For now, I think keeping an OR as a the default is best. But adapting the notation to have a comma mean "AND" may be easier since that's what a coma means elsewhere.